### PR TITLE
Add compiler validation for interceptors

### PR DESCRIPTION
**Pull request by TharmiganK**
## Purpose
$Subject

> Related to [`Introduce HTTP Filters #692`](https://github.com/ballerina-platform/ballerina-standard-library/issues/692)

## Examples
N/A

## Checklist
- [x] Linked to an issue
- [ ] <s>Updated the changelog</s>
- [x] Added tests

### CODE REVIEW

File: compiler-plugin/src/main/java/io/ballerina/stdlib/http/compiler/HttpInterceptorServiceValidator.java [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+public class HttpInterceptorServiceValidator implements AnalysisTask<SyntaxNodeAnalysisContext> {
+
+    @Override
+    public void perform(SyntaxNodeAnalysisContext syntaxNodeAnalysisContext) {
+        List<Diagnostic> diagnostics = syntaxNodeAnalysisContext.semanticModel().diagnostics();
+        for (Diagnostic diagnostic : diagnostics) {
+            if (diagnostic.diagnosticInfo().severity() == DiagnosticSeverity.ERROR) {
+                return;
+            }
+        }
```
Comment (ayeshLK) [inline compiler-plugin/src/main/java/io/ballerina/stdlib/http/compiler/HttpInterceptorServiceValidator.java:49]:
```suggestion
        boolean erroneousCompilation = diagnostics.stream()
                .anyMatch(d -> DiagnosticSeverity.ERROR.equals(d.diagnosticInfo().severity()));
        if (erroneousCompilation) {
            return;
        }
```

Comment (TharmiganK) [reply compiler-plugin/src/main/java/io/ballerina/stdlib/http/compiler/HttpInterceptorServiceValidator.java]: Fixed in [`042f3dc`](https://github.com/ballerina-platform/module-ballerina-http/pull/732/commits/042f3dcf2762bfe77eb39b7bed2aa415dd366f01)

File: compiler-plugin/src/main/java/io/ballerina/stdlib/http/compiler/HttpDiagnosticCodes.java [hunk 2]
```diff
@@ -62,6 +63,15 @@
...
     HTTP_120("HTTP_120", "invalid media-type subtype '%s'", ERROR),
     HTTP_121("HTTP_121", "invalid multiple 'http:RequestContext' parameter: '%s'", ERROR),
+    HTTP_122("HTTP_122", "invalid multiple 'error' parameter: '%s'", ERROR),
+    HTTP_123("HTTP_123", "invalid multiple interceptor type reference: '%s'", ERROR),
+    HTTP_124("HTTP_124", "invalid multiple interceptor resource functions", ERROR),
+    HTTP_125("HTTP_125", "invalid annotation '%s': annotations are not supported for interceptor " +
+            "resource functions", ERROR),
+    HTTP_126("HTTP_126", "invalid resource method return type: expected '" +
+              ALLOWED_INTERCEPTOR_RETURN_UNION + "', but found '%s'", ERROR),
+    HTTP_127("HTTP_127", "invalid resource path: expected default resource path, but found '%s'", ERROR),
```
Comment (chamil321) [inline compiler-plugin/src/main/java/io/ballerina/stdlib/http/compiler/HttpDiagnosticCodes.java:73]:
```suggestion
    HTTP_127("HTTP_127", "invalid interceptor resource path: expected default resource path: '[string... s]', but found '%s'", ERROR),
```

File: compiler-plugin/src/main/java/io/ballerina/stdlib/http/compiler/HttpDiagnosticCodes.java [hunk 3]
```diff
@@ -62,6 +63,15 @@
...
+    HTTP_128("HTTP_128", "invalid resource method: expected default resource method, but found '%s'", ERROR),
```
Comment (chamil321) [inline compiler-plugin/src/main/java/io/ballerina/stdlib/http/compiler/HttpDiagnosticCodes.java:74]:
```suggestion
    HTTP_128("HTTP_128", "invalid interceptor resource method: expected default resource method: 'default', but found '%s'", ERROR),
```

File: compiler-plugin/src/main/java/io/ballerina/stdlib/http/compiler/HttpDiagnosticCodes.java [hunk 4]
```diff
@@ -62,6 +63,15 @@
             ERROR),
     HTTP_120("HTTP_120", "invalid media-type subtype '%s'", ERROR),
     HTTP_121("HTTP_121", "invalid multiple 'http:RequestContext' parameter: '%s'", ERROR),
+    HTTP_122("HTTP_122", "invalid multiple 'error' parameter: '%s'", ERROR),
+    HTTP_123("HTTP_123", "invalid multiple interceptor type reference: '%s'", ERROR),
+    HTTP_124("HTTP_124", "invalid multiple interceptor resource functions", ERROR),
+    HTTP_125("HTTP_125", "invalid annotation '%s': annotations are not supported for interceptor " +
+            "resource functions", ERROR),
+    HTTP_126("HTTP_126", "invalid resource method return type: expected '" +
```
Comment (chamil321) [inline compiler-plugin/src/main/java/io/ballerina/stdlib/http/compiler/HttpDiagnosticCodes.java:71]:
```suggestion
    HTTP_126("HTTP_126", "invalid interceptor resource method return type: expected '" +
```

File: compiler-plugin/src/main/java/io/ballerina/stdlib/http/compiler/HttpInterceptorResourceValidator.java [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+        if (type.equals(Constants.REQUEST_ERROR_INTERCEPTOR)) {
+            extractAndValidateMethodAndPath(ctx, member);
+        }
+        HttpResourceValidator.extractInputParamTypeAndValidate(ctx, member);
+        extractReturnTypeAndValidate(ctx, member);
+    }
+
+    private static void extractAndValidateMethodAndPath(SyntaxNodeAnalysisContext ctx, FunctionDefinitionNode member) {
+        NodeList<Node> path = member.relativeResourcePath();
+        if (!path.isEmpty()) {
```
[the remark this answers is not in the document]
Comment (TharmiganK) [reply compiler-plugin/src/main/java/io/ballerina/stdlib/http/compiler/HttpInterceptorResourceValidator.java:54]: Yes we can ignore this

File: compiler-plugin/src/main/java/io/ballerina/stdlib/http/compiler/HttpInterceptorResourceValidator.java [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+            }
+        }
+        String method = member.functionName().toString().strip();
+        if (!method.contains(Constants.DEFAULT)) {
+            reportInvalidResourceMethod(ctx, member.functionName());
+        }
+    }
+
+    private static void checkResourceAnnotation(SyntaxNodeAnalysisContext ctx,
+                                           FunctionDefinitionNode member) {
```
[the remark this answers is not in the document]
Comment (TharmiganK) [reply compiler-plugin/src/main/java/io/ballerina/stdlib/http/compiler/HttpInterceptorResourceValidator.java]: Fixed in [`45d852c`](https://github.com/ballerina-platform/module-ballerina-http/pull/732/commits/45d852c3608119603f1bf9cd8616c27b76e0c2db)

File: compiler-plugin/src/main/java/io/ballerina/stdlib/http/compiler/HttpInterceptorResourceValidator.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (TharmiganK) [reply compiler-plugin/src/main/java/io/ballerina/stdlib/http/compiler/HttpInterceptorResourceValidator.java]: Fixed in [`45d852c`](https://github.com/ballerina-platform/module-ballerina-http/pull/732/commits/45d852c3608119603f1bf9cd8616c27b76e0c2db)

File: compiler-plugin/src/main/java/io/ballerina/stdlib/http/compiler/HttpDiagnosticCodes.java (hunk 4 above)
Comment (TharmiganK) [reply compiler-plugin/src/main/java/io/ballerina/stdlib/http/compiler/HttpDiagnosticCodes.java]: Fixed in [`45d852c`](https://github.com/ballerina-platform/module-ballerina-http/pull/732/commits/45d852c3608119603f1bf9cd8616c27b76e0c2db)

File: compiler-plugin/src/main/java/io/ballerina/stdlib/http/compiler/HttpDiagnosticCodes.java (hunk 3 above)
Comment (TharmiganK) [reply compiler-plugin/src/main/java/io/ballerina/stdlib/http/compiler/HttpDiagnosticCodes.java]: Fixed in [`45d852c`](https://github.com/ballerina-platform/module-ballerina-http/pull/732/commits/45d852c3608119603f1bf9cd8616c27b76e0c2db)

File: compiler-plugin/src/main/java/io/ballerina/stdlib/http/compiler/HttpDiagnosticCodes.java (hunk 2 above)
Comment (TharmiganK) [reply compiler-plugin/src/main/java/io/ballerina/stdlib/http/compiler/HttpDiagnosticCodes.java]: Fixed in [`45d852c`](https://github.com/ballerina-platform/module-ballerina-http/pull/732/commits/45d852c3608119603f1bf9cd8616c27b76e0c2db)

**Review by chamil321 — APPROVED** [review]

**Pull request merged**